x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java

```diff
@@ -54,10 +54,13 @@ protected InputStream openInputStream(final long position, final long length) th
                 @Override
                 protected InputStream openSlice(long slice) throws IOException {
                     final long currentPart = startPart + slice;
+                    final long startInPart = (currentPart == startPart) ? getRelativePositionInPart(position) : 0L;
+                    final long endInPart
+                        = (currentPart == endPart) ? getRelativePositionInPart(position + length) : getLengthOfPart(currentPart);
                     return blobContainer.readBlob(
                         fileInfo.partName(currentPart),
-                        (currentPart == startPart) ? getRelativePositionInPart(position) : 0L,
-                        (currentPart == endPart) ? getRelativePositionInPart(length) : getLengthOfPart(currentPart)
+                        startInPart,
+                        endInPart - startInPart
```

Comment:
Thanks, I was able to reproduce and came with a similar fix.

For better readability I'd suggest to isolate single part case (and be paranoid about the length to read) from the multiple parts case, something like:

```suggestion
        if (fileInfo.numberOfParts() == 1L) {
            return blobContainer.readBlob(fileInfo.partName(0L), position, Math.min(length, fileInfo.partBytes(0) - position));
        }

        final long startPart = getPartNumberForPosition(position);
        final long endPart = getPartNumberForPosition(position + length);

        return new SlicedInputStream(endPart - startPart + 1L) {
            @Override
            protected InputStream openSlice(long slice) throws IOException {
                final long currentPart = startPart + slice;
                final long startInPart = (currentPart == startPart) ? getRelativePositionInPart(position) : 0L;
                final long endInPart
                    = (currentPart == endPart) ? getRelativePositionInPart(position + length) : getLengthOfPart(currentPart);
                return blobContainer.readBlob(fileInfo.partName(currentPart), startInPart, endInPart - startInPart
                );
            }
        };
```

Reply:
Sure, sounds good, see 558e68a4d0f.